File: LeetCode/0994/golang/lc0994.go
package lc0994

import "fmt"

type Queue [][]int

func (q *Queue) Enqueue(val []int) {
	*q = append(*q, val)
}

func (q *Queue) Dequeue() ([]int, error) {
	if len(*q) == 0 {
		return []int{0, 0}, fmt.Errorf("Dequeue() from an empty Queue.")
	}
	val := (*q)[0]
	*q = (*q)[1:]
	return val, nil
}

func (q *Queue) Len() int {
	return len(*q)
}
func (q *Queue) Empty() bool {
	return len(*q) == 0
}

// Runtime complexity: O(HW)
// Auxiliary space: O(HW)
// where H, W are height and width of the `grid`.
func orangesRotting(grid [][]int) int {
	DIRS := [][]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}
	H := len(grid)
	W := len(grid[0])
	q := Queue{}
	numFreshOranges := 0

	for y := range H {
		for x := range W {
			if grid[y][x] == 1 {
				numFreshOranges += 1
			} else if grid[y][x] == 2 {
				q.Enqueue([]int{y, x})
			}
		}
	}
	if numFreshOranges == 0 {
		return 0
	}

	stepNum := 0
	for !q.Empty() {
		stepNum += 1
		for range q.Len() {
			startPos, _ := q.Dequeue()
			sy, sx := startPos[0], startPos[1]
			for _, dir := range DIRS {
				y := sy + dir[0]
				x := sx + dir[1]
				if y < 0 || y >= H || x < 0 || x >= W {
					continue
				}
				if grid[y][x] == 1 {
					grid[y][x] = 2
					q.Enqueue([]int{y, x})
					numFreshOranges--
				}
			}
		}
	}
	if numFreshOranges == 0 {
		return stepNum - 1
	}

	return -1
}

File: LeetCode/0994/python/lc0994.py
from typing import List
from collections import deque

# Runtime complexity: O(H*W)
# Auxiliary space complexity: O(H*W) in general case; depends on the number of cells enqueued.
# Subjective level: medium.
# Solved on: 2026-04-14
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        H = len(grid)
        W = len(grid[0])
        FIELD_ORANGE = 1
        FIELD_ROTTEN = 2
        DIRS = ((-1, 0), (0, 1), (1, 0), (0, -1))

        q = deque()
        totalFresh = 0
        for h in range(H):
            for w in range(W):
                if grid[h][w] == FIELD_ORANGE:
                    totalFresh += 1
                elif grid[h][w] == FIELD_ROTTEN:
                    q.append((h, w))

        # handle the corner case where there's no fresh oranges to begin with:
        if totalFresh == 0:
            return 0

        tick = 0
        while q: # in Python this means while !q.empty(). Can use this for any container. Nice!
            tick += 1
            for _ in range(len(q)):
                (h, w) = q.popleft() # note: want a FIFO queue? Need to call popleft(). Using pop() makes it a LIFO queue.
                for dir in DIRS:
                    nh, nw = h+dir[0], w+dir[1] # neighbor's position (nh, nw)
                    if nh < 0 or nh >= H or nw < 0 or nw >= W:
                        continue
                    # all the fresh oranges that neighbor a rotten one start rotting too:
                    if grid[nh][nw] == FIELD_ORANGE:
                        grid[nh][nw] = FIELD_ROTTEN
                        q.append((nh, nw))
                        totalFresh -= 1

        return -1 if totalFresh > 0 else tick-1
